lib/typescript-fast-incremental-preprocessor.js

```diff
@@ -0,0 +1,94 @@
+/* eslint-env node */
+const fs = require('fs');
+const path = require('path');
+const tmp = require('tmp')
+const child_process = require('child_process');
+const funnel = require('broccoli-funnel');
+const mergeTrees = require('broccoli-merge-trees');
+const Plugin = require('broccoli-plugin');
+
+class FastTypescriptWatcher extends Plugin {
+  constructor(inputNode, options = {}) {
+    super([ inputNode ], {
+      annotation: options.annotation,
+      needsCache: false,
+      name: 'ember-cli-typescript-fast-watcher',
+    });
+
+    this.outputDir = options.outputDir;
+    this.ui = options.ui;
+  }
+
+  build() {
+    if (this.watcher) {
+      return;
+    }
+    const inputDir = this.inputPaths[0];
+    this.watcher = child_process.fork(
+      require.resolve('typescript/bin/tsc'), [
+        '--watch',
+        '--project', inputDir,
+        '--outDir', this.outputDir,
+        '--noEmit', 'false'
+      ], { silent: true }); // TODO: stdout is messy
+
+    // TODO: clean up process
+  }
+}
+
+function makeTempDir() {
+  // TODO: the directory won't get monitored if it's in tmp/
+  const dir = '.e-c-ts';
+
+  if (!fs.existsSync(dir)){
+    fs.mkdirSync(dir);
+  }
+  return tmp.dirSync({ dir, prefix: 'ember_cli_typescript' });
+}
+
+class TypeScriptFastIncrementalPreprocessor {
+  constructor(options) {
+    this.name = 'ember-cli-typescript';
+    this.ui = options.ui;
+  }
+
+  toTree(inputNode, inputPath, outputPath) {
+    const js = funnel(inputNode, {
+      exclude: ['**/*.ts'],
+      annotation: 'JS files',
+    });
+
+    // TODO: how to disable watching .ts files?
```

Comment:
Once we figure out how to correctly identify the environment, it should either be available in the same way here *or* we can use it to set a flag on the `this` instance that we pass in so that we watch or not as appropriate for the environment.

Alternatively, maybe in the fast-compiler we intentionally just ignore these files entirely and only trigger on the files in the `.e-c-ts` directory changing?